### ISSUE TITLE
FileAppenderFactory not valid for null currentLogFilename

### ISSUE THREAD

**Issue by pandaadb** [dropwizard/dropwizard#1278]
Hey,

just noticed this. 

The currentLogFilename can be empty in case the RollingFileAppender is used. By default, this will result in the rollingFileAppender creating new names rather than using a static name. This is particularly helpful if one does not want renaming of files on rollover. 

Thanks,

Artur

**Comment by jplock** [comment]
Can you provide an example of this so we can add it to the documentation?

**Comment by pandaadb** [comment]
Hey,

yes sure, this is my factory extension: 

@JsonTypeName("hostnameFile")
public class HostnameFileAppender extends FileAppenderFactory implements AppenderFactory {
	
	private static final String HACKERY = "NO_VALUE_DROPWIZARD_HACKERY";
	
	public HostnameFileAppender() {
		setCurrentLogFilename(HACKERY); // hackery to make validation work.
	}
	@Override
	protected FileAppender<ILoggingEvent> buildAppender(LoggerContext context) {
		RollingFileAppender<ILoggingEvent> appender =  (RollingFileAppender<ILoggingEvent>) super.buildAppender(context);
		if(HACKERY.equals(getCurrentLogFilename())) {
			appender.setFile(null); // set this name to null if the hackery is applied. We want this, it means there is no renaming of log files going on.
		}
		return appender;
	}
	
}

Note the construct setting the value of the file name. This is because currentLogFilename in FileAppenderFactory is annoted with a @NotNull annotation. 

In the create method  it resets the filename to Null. This enabled the default FileAppender to use the archive name instead. It will then create file names with the rollovername by default, not renaming any of the files when rolling over. 

In our use case this was important because we rsync logs to our logstash deployment, and we a rename of the current log file would cause double parsing which we don't want. 

Note: you will have to enable the new factory by adding it to the services in META-INF/services so that it can be used in the yaml file. 

Hope this helps, let me know if you need more info.

**Comment by nickbabcock** [comment]
@pandaadb, I think it may only be for certain rolling policies (correct me if I'm wrong) that `setFile(null)` will work as if `getFile()` is null in `FileAppender` the appender won't start.

While I want Dropwizard users not to be unnecessarily constrained by validation annotations, I'd hate to create more ways for users to create invalid configs. I am open for persuasion, though. Thoughts?

**Comment by pandaadb** [comment]
@nickbabcock 

I have looked through it and you are right. My "hack" only works for the correct appender since the default FileAppender does not like to have null values. 

Thing is, even the RollingFileAppender explodes with the file being null. Maybe this can be a different validation. I looked through your factory and it appears that an elegant solution might be something like that: 

 ```
   @ValidationMethod(message = "File must not be null")
    public boolean isValidArchiveConfiguration() {
        return getFile() != null;
    }
```

This way, instead of validating the actual property, it validates the functionality instead. Since for the rolling appenders, the policy defines the filename. 

I hope that makes sense?

**Comment by nickbabcock** [comment]
Great point -- your solution makes total sense. Please submit a PR with it and some tests.

And the validation message should be along the lines of:

> currentLogFilename can only be null when archiving is enabled

If feel unable to write the PR just let us know.

**Comment by pandaadb** [comment]
Hey - first time pull request. Let me know if I messed it up and/or something is missing :) That was much more convenient that I thought it would be. Might do that more often :)

**Comment by nickbabcock** [comment]
Closed by #1278
